Introduced interfaces to remove viper dependencies from the config pa…

manishSan opened the pull request:
Introduced interfaces to remove viper dependencies from the config package.
Next step is to write test with Mock.

Commit:
Introduced interfaces to remove viper dependencies from the config package.

Commit:
Merge branch 'master' into feature/remove_config_dependencies

Fixed merge conflict.

superarts commented on config/configurator.go:
```diff
@@ -0,0 +1,44 @@
+	configurationReader
+}
+
+type ConfigProvider struct {
```
From my understanding, we are injecting `ConfigProvider` into `Configuration`. `ConfigProvider` strongly depends on vendor library i.e. `viper`, and potentially can be replaced with something else.

If the assumption above is right, I have 2 questions:

1. Shall we rename `ConfigProvider` to `ViperProvider`, to specifically identify vendor dependency?
2. Is it the preferred way to handle dependency injection at the caller level, i.e. `config.Init(new(config.ViperProvider))`?

I have no problem with constructor injection, but if we follow the diagram and whenever we want to initialize a new component, we will have to follow the `package.Init(new(package.WhateverProvider))` pattern, I'm not sure if it's ideal. I know under the current design we can use:

- `package.Init(new(package.VendorProvider))` for "real" implementation,
- `package.Init(new(package.MockProvider))` for unit testing;

We can also create helper functions to use:

- `package.VendorSetup()` for "real" implementation,
- `package.MockSetup()` for unit testing;

And of course, we can also use DI container like [this](https://github.com/sarulabs/di). Personally I would prefer creating helper functions before introducing DI container, e.g. `config` initialization will look like:

- `config.ViperSetup()` or `config.SetupWithViper()` or something

I know you may not be a big fan of wrapper functions, I can compromise if you do insist. But I would at least writing the code in one line, as `provider` is only used in `Init` and an additional variable for each package initializer makes me feel a little big uneasy.

manishb24 replied:
Renamed `ConfigProvider` to `ViperProvider`.
at some point in future I'll add DI container and would pass container to the constructor instead of independent dependencies.

superarts replied:
Cool let's go with the current code for now, and whenever we are going to work on refactoring the 2nd component, let's do either the 2nd or 3rd way together.

superarts commented on config/configurator.go:
```diff
@@ -0,0 +1,44 @@
+	merge() error
+}
+
+type configurationReader interface {
```
As it's inside `config` package, I think just `Reader` and `Loader` would be enough.

manishb24 replied:
Done.

superarts commented on config/configurator.go:
```diff
@@ -0,0 +1,44 @@
+type configurationLoader interface {
+	setConfig(name string, extension string, path string)
+	read() error
+	merge() error
```
I'm not sure about `merge` - to me it's something related with the design of `viper`, for other vendor packages(s) there may not be such concept. I would say this whole `setName -> read -> setAnotherName -> merge` thing is viper specific, and should be handled in a different layer.

superarts replied:
I think we can put `setConfig`, `read`, and `merge` together to `loadConfig(filename string)`, because from the high level view, what `config` needs to do is to convert config files into native struct. Low level provider should take a filename, and parse and persist the actual config key/value pairs (to allow `getBool` and `getString` later), and should allow high level caller to do:

- load -> load another -> load another ...
- possibly provide a "clear" function, which is not a business requirement now

i.e. `merge` should be the default behavior as that's the usual use case of server side config.

superarts commented on config/configurator.go:
```diff
@@ -0,0 +1,44 @@
+
+import "github.com/spf13/viper"
+
+type configurationLoader interface {
```
Minor naming thing: I know "Loader" handles environment setup and it seems to be fine, but "Reader" actually "loads" configs from file to struct, so "Reader" sounds more like a "Loader". But in this case I'm not sure what we should call "Loader" - both "Initializer" and "Configer" sound weird to me. But again, naming is minor.

manishb24 replied:
This is for future discussion.

superarts reviewed (commented):
Check my comments please.

Commit:
Merge branch 'master' into feature/remove_config_dependencies

Commit:
Review comments -
	renamed ConfigProvider -> ViperProvider
	renamed ConfigLoader -> Loader
	renamed ConfigReader -> Reader

manishb24 commented on .gitignore:
```diff
@@ -1,5 +1,7 @@
 .DS_Store
 .*.sw*
 hindsight
+debug
```
yup.

manishb24 commented on config/config.go:
```diff
@@ -1,69 +1,61 @@
 
 	JWT_Realm string
-	JWT_Key string
+	JWT_Key   string
```
Yes, VsCode does this formatting automatically.

superarts reviewed (commented):
Only one thing left
